pkg/pagination/pagination.go

```
@@ -0,0 +1,103 @@
+// being used to keep track of pages.
+type Key string
+
+// MustGenerateNewKey generates a new pagination key.
+//
+// This function will panic if your system does not have enough entropy to
+// produce the required randomness.
+func MustGenerateNewKey() Key {
+	var key fernet.Key
+	if err := key.Generate(); err != nil {
+		panic("pagination: failed to read enough randomness to generate a new key")
+	}
+
+	return Key(key.Encode())
+}
+
+// KeyFromString creates the key for a given string.
+//
+// Strings must be 32-bit URL-safe base64 representations of the key bytes.
+func KeyFromString(keyString string) (k Key, err error) {
+	var key *fernet.Key
+	key, err = fernet.DecodeKey(keyString)
+	if err != nil {
+		err = errors.New("invalid pagination key string: must be 32-bit URL-safe base64")
+	}
+	return Key(key.Encode()), err
+}
+
+// MustKeyFromString creates the key for a given string.
+//
+// This function will panic if the key is not well-formed.
+//
+// You may find this function useful if you have already called KeyFromString
+// and know the provided value is safe.
+func MustKeyFromString(keyString string) Key {
+	k, err := KeyFromString(keyString)
+	if err != nil {
+		panic("pagination: failed to encode string as key")
+	}
+
+	return k
+}
+
+// String implements the fmt.Stringer interface for Key.
+func (k Key) String() string {
+	return string(k)
+}
+
+// MarshalToken encodes an interface into JSON bytes and encrypts it.
+func (k Key) MarshalToken(v interface{}) ([]byte, error) {
+	var buf bytes.Buffer
+	err := json.NewEncoder(&buf).Encode(v)
+	if err != nil {
+		return nil, err
+	}
+
+	decoded, _ := fernet.DecodeKey(string(k))
+	return fernet.EncryptAndSign(buf.Bytes(), decoded)
+}
+
+// UnmarshalToken decrypts a token using provided key and decodes the result
+// into the provided interface.
+func (k Key) UnmarshalToken(token string, v interface{}) error {
+	decoded, _ := fernet.DecodeKey(string(k))
```

Do we really want to suppress the error here? Yes, it was not handled before. If there's a valid reason, I think it makes sense to comment it.

---

If you only ever make Keys from the constructors, they'd always be valid.

However, I'll change this so that the Key type actually keeps the decoded bytes in memory rather than the string representation.